include/lsst/afw/geom/SkyWcs.h

```diff
@@ -94,158 +100,391 @@ Constructors must set the following properties of the contained ast::FrameSet:
+                                         AngleUnit const &skyUnit) const;
+
+    /// Compute _pixelOrigin and _pixelScaleAtOrigin
+    void _computeCache() {
+        _pixelOrigin = skyToPixel(getSkyOrigin());
+        _pixelScaleAtOrigin = getPixelScale(_pixelOrigin);
+    }
 
-    // Check a FrameSet to see if it can safely be used for a SkyWcs
-    // Return a copy so that it can be used as an argument to the SkyWcs(shared_ptr<FrameSet>) constructor
-    std::shared_ptr<ast::FrameSet> _checkFrameSet(ast::FrameSet const &frameSet) const;
 };
 
+/**
+ * Return a copy of a FITS-WCS with pixel positions flipped around a specified center
+ *
+ * @param[in] wcs  The initial WCS
+ * @param[in] flipLR  Flip pixel positions left/right about `center`
+ * @param[in] flipTB  Flip pixel positions top/bottom about `center`
+ * @param[in] center  Center pixel position of flip
+ */
+std::shared_ptr<SkyWcs> makeFlippedWcs(SkyWcs const &wcs, bool flipLR, bool flipTB,
+                                       geom::Point2D const &center);
+
+/**
+ * Create a new SkyWcs whose pixels are transformed by pixelTransform, as described below.
+ *
+ * If modifyActualPixels is true and the ACTUAL_PIXELS frame exists then pixelTransform
+ * is inserted just after the ACTUAL_PIXELS frame:
+ *    newActualPixelsToPixels = pixelTransform -> oldActualPixelsToPixels
+ * This is appropriate for shitting a WCS, e.g. when writing FITS metadata for a subimage.
+ *
+ * If modifyActualPixels is false or the ACTUAL_PIXELS frame does not exist then pixelTransform
+ * is inserted just after the PIXELS frame:
+ *    newPixelsToIwc = pixelTransform -> oldPixelsToIwc
+ * This is appropriate for inserting a model for optical distortion.
+ *
+ * Other than the change described above, the new SkyWcs will be just like the old one.
+ *
+ * @param[in] pixelTransform  Transform to insert
+ * param[in] modifyActualPixels  Location at which to insert the transform;
+ *    if true and the ACTUAL_PIXELS frame is present then insert just after the ACTUAL_PIXELS frame,
+ *    else insert just after the PIXELS frame.
+ */
+std::shared_ptr<SkyWcs> makeModifiedWcs(TransformPoint2ToPoint2 const &pixelTransform, SkyWcs const &wcs,
+                                        bool modifyActualPixels);
+
+/**
+ *  Construct a TAN-SIP WCS
+ *
+ * Construct a TAN WCS with forward SIP distortion terms and an iterative inverse.
+ *
+ * @param[in] crpix    Center of projection on the CCD using the LSST convention:
+ *                      0, 0 is the lower left pixel of the parent image
+ * @param[in] crval    Center of projection on the sky
+ * @param[in] cdMatrix CD matrix, where element (i-1, j-1) corresponds to FITS keyword CDi_j
+ *                      and i, j have range [1, 2]. May be computed by calling makeCdMatrix.
+ * @param[in] sipA     Forward distortion matrix for axis 1
+ * @param[in] sipB     Forward distortion matrix for axis 2
+ */
```

Comment:
`crpix` and `crval` are so widely used that I felt it would be too disruptive to change it.

I can use `cd` instead of `cdMatrix` if you strongly prefer that; I dislike it because it is so short as to be almost meaningless.

It is all a bit of a mess, since both the old `Wcs` and new `SkyWcs` get `crpix` by calling `getPixelOrigin` and `crval` by calling `getSkyOrigin`. I tried to clean it up a bit by unifying `cdMatrix` (which was rarely used because `makeWcs` took it as 4 floats, not by name).

Trying to change all existing usage of crpix to pixelOrigin and crval to skyOrigin is likely to be very unpopular due to the widespread usage.